### PR TITLE
Add Addressable Pixel API

### PR DESCRIPTION
**Pull request by brentru**
Adds the ability for a WipperSnapper device to interface with a strand (size n) of addressable RGB Pixels (DotStar, NeoPixel).

**Note**: This PR only supports setting an entire strand to one-color ("filling a strand"). WipperSnapper does not yet support individually addressing each pixel or pixel animation/sequence.

Address https://github.com/adafruit/Wippersnapper_Protobuf/issues/90

### CODE REVIEW

File: src/components/pixels/ws_pixels.cpp [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+/**************************************************************************/
+/*!
+    @brief  Destructor
+*/
+/**************************************************************************/
+ws_pixels::~ws_pixels() {
+  // release all strands
+  for (int i = 0; i < sizeof(_strands); i++)
+    deallocateStrand(i);
+  // re-init the status LED
```
Comment (brentru) [inline src/components/pixels/ws_pixels.cpp:45]: We may want to check if this was in-use previously..

File: src/components/pixels/ws_pixels.cpp [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+  case wippersnapper_pixels_v1_PixelsOrder_PIXELS_ORDER_RGB:
+    strandType = NEO_RGB + NEO_KHZ800;
+    break;
+  case wippersnapper_pixels_v1_PixelsOrder_PIXELS_ORDER_RGBW:
+    strandType = NEO_RGBW + NEO_KHZ800;
+    break;
+  case wippersnapper_pixels_v1_PixelsOrder_PIXELS_ORDER_BRG:
+    strandType = NEO_BRG + NEO_KHZ800;
+    break;
+  default:
```
Comment (brentru) [inline src/components/pixels/ws_pixels.cpp:113]: Set `strandType = NEO_GRB + NEO_KHZ800` as default here, remove declaration on L96.

File: src/components/pixels/ws_pixels.cpp [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+    _strands[strandIdx].pinDotStarClock = atoi(pinClock);
+    // release the status dotstar, if it is both in-use and the pin within
+    // `pixelsCreateReqMsg`
+    if ((_strands[strandIdx].pinDotStarData == getStatusDotStarDataPin()) &&
+        WS.lockStatusDotStar)
+      releaseStatusLED();
+    // init DotStar object
+    _strands[strandIdx].dotStarPtr = new Adafruit_DotStar(
+        pixelsCreateReqMsg->pixels_num, _strands[strandIdx].pinDotStarData,
+        _strands[strandIdx].pinDotStarClock,
```
Comment (brentru) [inline src/components/pixels/ws_pixels.cpp:214]: Call to `getDotStarStrandOrder` should handle case of `-1`

File: src/components/pixels/ws_pixels.cpp [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+/*!
+    @brief   Obtains the index of a `strand_t` within array of `_strands`.
+    @param   dataPin
+             strand_t's data dataPin
+    @param   type
+             Type of strand_t, NeoPixel or DotStar.
+    @returns The index of a strand_t if within strands[], -1 otherwise.
+*/
+/**************************************************************************/
+int ws_pixels::getStrandIdx(int16_t dataPin,
```
Comment (brentru) [inline src/components/pixels/ws_pixels.cpp:271]: Do we replicate this anywhere else in the code, can we call getStrandIdx instead?

File: src/components/pixels/ws_pixels.cpp [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+  if (strandIdx == -1) {
+    WS_DEBUG_PRINTLN(
+        "ERROR: Strand not found, can not write a color to the strand!");
+    return;
+  }
+
+  if (pixelsWriteMsg->pixels_type ==
+          wippersnapper_pixels_v1_PixelsType_PIXELS_TYPE_NEOPIXEL) {
+    // let's fill the strand
+    uint16_t numPixels = _strands[strandIdx].neoPixelPtr->numPixels();
```
Comment (brentru) [inline src/components/pixels/ws_pixels.cpp:357]: call `_strands[strandIdx].neoPixelPtr->numPixels();` inside loop instead

File: src/components/pixels/ws_pixels.h [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+ *
+ * MIT license, all text here must be included in any redistribution.
+ *
+ */
+#ifndef WS_PIXELS
+#define WS_PIXELS
+
+#include "Wippersnapper.h"
+
+#define MAX_PIXEL_STRANDS                                                      \
```
Comment (brentru) [inline src/components/pixels/ws_pixels.h:22]: Do we enforce this anywhere within `ws_pixels.cpp`?

**Comment by brentru** [comment]
@caternuson  Could you please review the `ws_pixels` class and code? Thanks in advance!

**Comment by caternuson** [comment]
Gave the code a *very* cursory look, only looking at the NeoPixel/DotStar library calls being done. The setup calls done in `ws_pixels::addStrand()` look OK. The fill calls in `ws_pixels::fillStrand()` look OK.

**Pull request merged**